### PR TITLE
feat(functional): make the functional api client a pure function.

### LINKED ISSUE #97: Do you have the plan to suppor tree-shaking?
I'm experimenting this code generator. It's functionality is meet to my use-case.

However, I have one problem to use this. The generated code is big one class so if I try to use one of the API in the client, then all api's are bundled.

Do you have the plan to improve this?

### PR DESCRIPTION
**Pull request by hrsh7th**
## Summary

Currently, the functional api client does not support tree-shaking.
This PR solves the problem.

Fixes #97 

## Test Plan

I'll update the snapshot testing.

## Example output

```ts
...

export const RolloutService_SetRolloutImage = <RequestOption>(apiClient: ApiClient<RequestOption>, params: Params$RolloutService_SetRolloutImage, option?: RequestOption): Promise<Response$RolloutService_SetRolloutImage$Status$200["application/json"]> => {
    const url = `/api/v1/rollouts/`;
    const headers = {
        "Content-Type": "application/json",
        Accept: "application/json"
    };
    return apiClient.request("PUT", url, headers, params.requestBody, undefined, option);
};
export const RolloutService_UndoRollout = <RequestOption>(apiClient: ApiClient<RequestOption>, params: Params$RolloutService_UndoRollout, option?: RequestOption): Promise<Response$RolloutService_UndoRollout$Status$200["application/json"]> => {
    const url = `/api/v1/rollouts/`;
    const headers = {
        "Content-Type": "application/json",
        Accept: "application/json"
    };
    return apiClient.request("PUT", url, headers, params.requestBody, undefined, option);
};
export const RolloutService_Version = <RequestOption>(apiClient: ApiClient<RequestOption>, option?: RequestOption): Promise<Response$RolloutService_Version$Status$200["application/json"]> => {
    const url = `/api/v1/version`;
    const headers = {
        Accept: "application/json"
    };
    return apiClient.request("GET", url, headers, undefined, undefined, option);
};

...
```

I don't know why but the `url` variable is corrupted now...

### CODE REVIEW

File: src/code-templates/_shared/MethodBody/PathParameter.ts [hunk 1]
```diff
@@ -18,25 +18,23 @@ const generateUrlVariableStatement = (
...
+  const expression = {
+    class: factory.BinaryExpression.create({
+      left: factory.PropertyAccessExpression.create({
+        name: "baseUrl",
+        expression: "this",
+      }),
+      operator: "+",
+      right: Utils.generateTemplateExpression(factory, urlTemplate),
     }),
+    function: Utils.generateTemplateExpression(factory, urlTemplate),
```
Comment (hrsh7th) [inline src/code-templates/_shared/MethodBody/PathParameter.ts:30]: I can't understand the reason but this code isn't worked.

The `class` part works fine but `function` part will be wrong template string.

File: src/code-templates/functional-api-client/FunctionalApiClient/ArrowFunction.ts [hunk 2]
```diff
@@ -121,14 +121,35 @@ const methodTypeParameters = (factory: TsGenerator.Factory.Type, { convertedPara
...
 export const create = (factory: TsGenerator.Factory.Type, params: CodeGenerator.Params, option: Option): ts.ArrowFunction => {
   const { convertedParams } = params;
-  const typeParameters: ts.TypeParameterDeclaration[] = methodTypeParameters(factory, params);
+  const typeParameters: ts.TypeParameterDeclaration[] = [];
+  typeParameters.push(
+    factory.TypeParameterDeclaration.create({
+      name: "RequestOption",
+    }),
+  );
+  typeParameters.push(...methodTypeParameters(factory, params));
```
Comment (hrsh7th) [inline src/code-templates/functional-api-client/FunctionalApiClient/ArrowFunction.ts:130]: each functions are now accept the `apiClient` argument and `RequestOption` type argument.

File: src/code-templates/functional-api-client/FunctionalApiClient/ArrowFunction.ts [hunk 3]
```diff
@@ -121,14 +121,35 @@ const methodTypeParameters = (factory: TsGenerator.Factory.Type, { convertedPara
...
+      type: factory.TypeReferenceNode.create({
+        name: "ApiClient",
+        typeArguments: [
+          factory.TypeReferenceNode.create({
+            name: "RequestOption",
+          }),
+        ],
+      }),
+    }),
+  );
```
Comment (hrsh7th) [inline src/code-templates/functional-api-client/FunctionalApiClient/ArrowFunction.ts:146]: the `apiClient` argument was added here.

File: src/code-templates/functional-api-client/FunctionalApiClient/index.ts [hunk 4]
```diff
@@ -1,89 +1,23 @@
...
+export const create = (factory: TsGenerator.Factory.Type, list: CodeGenerator.Params[], option: Option): ts.VariableStatement[] => {
+  return list.map(params => {
+    return factory.VariableStatement.create({
+      modifiers: [ts.factory.createToken(ts.SyntaxKind.ExportKeyword)],
+      declarationList: factory.VariableDeclarationList.create({
+        declarations: [
+          factory.VariableDeclaration.create({
+            name: params.convertedParams.functionName,
+            initializer: ArrowFunction.create(factory, params, option),
           }),
```
Comment (hrsh7th) [inline src/code-templates/functional-api-client/FunctionalApiClient/index.ts:17]: the functional api client will be `list of ArrowFunctions` now.

**Comment by Himenon** [comment]
@hrsh7th 
Thank you for your cooperation. We have created #109 based on this idea, which has been committed as Co-authored, so please check it out.

**Comment by hrsh7th** [comment]
Fixed by #109. Closing.

**Pull request closed without merging**